### PR TITLE
The `bool` type is neither a host-shareable nor an interface type.

### PR DESCRIPTION
**Pull request by jimblandy**
However, `bool` is permitted for built-ins.

### CODE REVIEW

**Comment by jimblandy** [comment]
Actually, it might be best to combine this with a fix for `TypeFlags::INTERFACE` handling.

File: src/valid/type.rs [hunk 1]
```diff
@@ -332,7 +335,12 @@ impl super::Validator {
                         return Err(TypeError::InvalidData(member.ty));
                     }
                     if top_level && !base_info.flags.contains(TypeFlags::INTERFACE) {
-                        return Err(TypeError::InvalidBlockType(member.ty));
+                        // Certain non-`INTERFACE` types like arrays and `bool` are
+                        // permitted for built-ins, whose types we verify in
+                        // `VaryingContext::validate`.
+                        if !matches!(member.binding, Some(crate::Binding::BuiltIn(_))) {
```
Comment (kvark) [inline src/valid/type.rs:341]: let's not use matches please

Comment (jimblandy) [reply src/valid/type.rs]: Here, `matches!` doesn't buy much, but there's another spot I ran into (in another patch series) where the `match` is worse. I'll point it out when we get there, curious what you'll think.

**Review by kvark — APPROVED** [review]

**Comment by teoxoy** [comment]
Can we merge this?

**Comment by jimblandy** [comment]
It's probably worth re-reviewing, since it's so old.

**Comment by teoxoy** [comment]
I can re-review it but I think it needs to be rebased first.

**Comment by jimblandy** [comment]
Okay - this version is basically a rewrite for the current sources. It depends on #1875 (and thus is still a draft), and conflicts with #1874. If we can get those landed, I'll re-base and move this out of "draft" state.

**Comment by jimblandy** [comment]
This has some renames and some refactoring, so it's probably easiest to review commit-by-commit.

Each commit should pass CI individually, so it's not necessary to squash before merging.

**Review by teoxoy — APPROVED** [review]
Looks good but the title indicates that the PR will also remove `bool`s from being `HOST_SHAREABLE`.

**Review by teoxoy — APPROVED** [review]
Looks great!

**Pull request merged**